pkg/funky/router.go

```diff
@@ -0,0 +1,199 @@
+		servers: servers,
+		mutex:   &sync.Mutex{},
+		sem:     semaphore.NewWeighted(int64(numServers)),
+	}, nil
+}
+
+// Delegate delegates function invocation to an idle server
+func (r *DefaultRouter) Delegate(input Message) (*Message, error) {
+	server, err := r.findFreeServer()
+	if err != nil {
+		return nil, err
+	}
+
+	defer r.releaseServer(server)
+
+	var e Error
+	resp, err := server.Invoke(input)
+	if resp != nil {
+		defer resp.Close()
+	}
+	if err != nil {
+		switch err.(type) {
+		case TimeoutError:
+			e = Error{
+				ErrorType: FunctionError,
+				Message:   err.Error(),
+			}
+			break
+		case BadRequestError:
+			e = Error{
+				ErrorType: InputError,
+				Message:   err.Error(),
+			}
+			break
+		case InvocationError:
+			e = Error{
+				ErrorType: FunctionError,
+				Message:   err.Error(),
+			}
+			break
+		case InvalidResponsePayloadError:
+			e = Error{
+				ErrorType: InputError,
+				Message:   err.Error(),
+			}
+			break
+		case UnknownSystemError:
+			e = Error{
+				ErrorType: SystemError,
+				Message:   err.Error(),
+			}
+			break
+		default:
+			e = Error{
+				ErrorType: SystemError,
+				Message:   err.Error(),
+			}
+			break
+		}
+	}
+
+	logs := Logs{
+		Stdout: splitLogsOnNewline(server.Stdout()),
+		Stderr: splitLogsOnNewline(server.Stderr()),
+	}
+
+	context := Context{
+		Logs:  &logs,
+		Error: &e,
+	}
+
+	respBuf := new(bytes.Buffer)
+	if resp != nil {
+		respBuf.ReadFrom(resp)
+	}
+	respPayload := make(map[string]interface{})
+	json.Unmarshal(respBuf.Bytes(), &respPayload)
+
+	response := &Message{
+		Context: &context,
+		Payload: respPayload,
+	}
+
+	return response, nil
+}
+
+// Shutdown shuts down the servers managed by this router
+func (r *DefaultRouter) Shutdown() error {
+	var err error
+	for _, server := range r.servers {
+		err = server.Shutdown()
+	}
+
+	if err != nil {
+		return errors.New("Failed to shutdown one or more servers")
+	}
+
+	return nil
+}
+
+func createServers(numServers int, serverFactory ServerFactory) ([]Server, error) {
+	servers := []Server{}
+	for i := uint16(0); i < uint16(numServers); i++ {
+		server, err := serverFactory.CreateServer(FirstPort + i)
+		if err != nil {
+			return nil, fmt.Errorf("Failed creating server on port %d", FirstPort+i)
+		}
+		if err := server.Start(); err != nil {
+			return nil, fmt.Errorf("Failed to start server %+v", err)
+		}
+		servers = append(servers, server)
+	}
+
+	return servers, nil
+}
+
+func (r *DefaultRouter) findFreeServer() (Server, error) {
+	var ret Server
+
+	r.sem.Acquire(context.TODO(), 1)
+	for _, server := range r.servers {
+		r.mutex.Lock()
+		if server.IsIdle() {
+			server.SetIdle(false)
+			ret = server
+		}
+		r.mutex.Unlock()
+		if ret != nil {
+			break
+		}
+	}
+
+	if ret == nil {
+		r.sem.Release(1)
+		return nil, errors.New("no free server")
+	}
+
+	return ret, nil
+}
+
+func (r *DefaultRouter) releaseServer(server Server) {
+	r.mutex.Lock()
+	server.SetIdle(true)
+	r.mutex.Unlock()
+	r.sem.Release(1)
+}
```

Comment:
`findFreeServer()` and `releaseServer(server Server)` could be simplified, e.g.:

```go
func (r *DefaultRouter) findFreeServer() (Server, error) {
	if err := r.sem.Acquire(context.TODO(), 1); err != nil {
		return nil, err
	}

	// if we're here, it's guaranteed we have at least one element in servers

	r.mutex.Lock()
	defer r.mutex.Unlock()

	server := r.servers[0]
	r.servers = r.servers[1:]

	return server, nil
}

func (r *DefaultRouter) releaseServer(server Server) {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	r.servers = append(r.servers, server)
	r.sem.Release(1)
}
```